File: app/src/main/java/online/taxcore/pos/utils/CreatePdf.java
package online.taxcore.pos.utils;

import android.os.Environment;

import com.itextpdf.text.BaseColor;
import com.itextpdf.text.Document;
import com.itextpdf.text.DocumentException;
import com.itextpdf.text.Element;
import com.itextpdf.text.Font;
import com.itextpdf.text.Image;
import com.itextpdf.text.Paragraph;
import com.itextpdf.text.Rectangle;
import com.itextpdf.text.pdf.PdfPCell;
import com.itextpdf.text.pdf.PdfPTable;
import com.itextpdf.text.pdf.PdfWriter;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

public class CreatePdf {
    //write method takes two parameter pdf name and content
    //return true if pdf successfully created
    public static Boolean write(String fname, String fcontent, byte[] imageByteArray, String pdf_footer) {
        try {
            //Create file path for Pdf
            String fpath = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).toString() + "/" + fname + ".pdf";
            File file = new File(fpath);
            if (!file.exists()) {
                file.createNewFile();
            }
            // To customise the text of the pdf
            // we can use FontFamily
            Font bfBold12 = new Font(Font.FontFamily.COURIER,
                    7, Font.BOLD, new BaseColor(0, 0, 0));

            Image instance = Image.getInstance(imageByteArray);
            instance.scaleAbsolute(168, 168);

            PdfPTable table = new PdfPTable(1);
            table.setTotalWidth(174);
            table.setLockedWidth(true);
            PdfPCell cell = new PdfPCell();
            cell.setHorizontalAlignment(Element.ALIGN_CENTER);

            cell.setBorder(Rectangle.NO_BORDER);
            Paragraph p1 = new Paragraph(fcontent, bfBold12);
            p1.setLeading(0, 1);
            cell.addElement(p1);

            cell.addElement(instance);

            Paragraph p2 = new Paragraph(pdf_footer, bfBold12);
            p2.setLeading(0, 1);
            cell.addElement(p2);

            table.addCell(cell);

            Rectangle pagesize = new Rectangle(174, 800);
            Document document = new Document(pagesize);
            document.setMargins(5, 5, 0, 0);

            PdfWriter.getInstance(document, new FileOutputStream(file.getAbsoluteFile()));
            document.open();

            document.add(table);

            // close document
            document.close();
            return true;
        } catch (IOException | DocumentException e) {
            e.printStackTrace();
            return false;
        }
    }
}

File: app/src/main/java/online/taxcore/pos/utils/CreatePdf.kt
// TODO: Migrate to iText7 - This file needs to be completely rewritten for iText7
// Commenting out for now to avoid build errors
/*
package online.taxcore.pos.utils

import com.itextpdf.kernel.colors.Color
import com.itextpdf.kernel.colors.DeviceRgb
import com.itextpdf.kernel.font.PdfFont
import com.itextpdf.kernel.font.PdfFontFactory
import com.itextpdf.kernel.geom.Rectangle
import com.itextpdf.kernel.pdf.PdfDocument
import com.itextpdf.kernel.pdf.PdfWriter
import com.itextpdf.layout.Document
import com.itextpdf.layout.element.Image
import com.itextpdf.layout.element.Paragraph
import com.itextpdf.layout.element.Table
import com.itextpdf.layout.property.TextAlignment
import com.itextpdf.layout.property.VerticalAlignment
import com.itextpdf.io.image.ImageDataFactory
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

object CreatePdf {
    private const val FONT_CONSOLA_MONO = "/assets/fonts/ConsolaMono.ttf"

    //write method takes two parameter pdf name and content
    //return true if pdf successfully created
    fun write(
        filePath: String?,
        invoiceJournal: String,
        imageByteArray: ByteArray?
    ): Boolean {
        return try {
            val file = File(filePath.toString())
            if (!file.exists()) {
                file.createNewFile()
            }

            val delimiter = "========================================\r\n"
            val invoiceHeader = invoiceJournal.split(delimiter).first()

            val headerElements = invoiceHeader.split("\r\n".toRegex())

            val headerStart = headerElements.first()
            val companyHeader = headerElements.subList(1, 6).joinToString("\r\n")
            val headerInfo = headerElements.subList(6, headerElements.size - 1).joinToString("\r\n")

            val invoiceFooter = invoiceJournal.split(delimiter).last()
            val invoiceMain = invoiceJournal
                .replace(invoiceHeader, "")
                .replace(invoiceFooter, "")

            val font = PdfFontFactory.createFont(FONT_CONSOLA_MONO, "Identity-H", PdfFontFactory.EmbeddingStrategy.FORCE_EMBEDDED)
            val qrCodeImage = Image(ImageDataFactory.create(imageByteArray))
            qrCodeImage.scaleAbsolute(163f, 163f)

            val pHeaderStart = Paragraph(headerStart).setFont(font).setFontSize(7f)
            pHeaderStart.setLeading(0f, 1f)

            val pCompanyText = Paragraph(companyHeader).setFont(font).setFontSize(7f)
            pCompanyText.setLeading(0f, 1f)
            pCompanyText.alignment = Element.ALIGN_CENTER

            val pHeaderInfo = Paragraph(headerInfo, bfBold12)
            pHeaderInfo.setLeading(0f, 1f)

            val pInvoiceMain = Paragraph(invoiceMain, bfBold12)
            pInvoiceMain.setLeading(0f, 1f)

            val pFooterText = Paragraph(invoiceFooter, bfBold12)
            pFooterText.setLeading(0f, 1f)

            val cell = PdfPCell()
            cell.border = Rectangle.NO_BORDER

            cell.addElement(pHeaderStart)
            cell.addElement(pCompanyText)
            cell.addElement(pHeaderInfo)
            cell.addElement(pInvoiceMain)
            cell.addElement(qrCodeImage)
            cell.addElement(pFooterText)

            val table = PdfPTable(1)
            table.totalWidth = 174f
            table.isLockedWidth = true
            table.addCell(cell)

            val rows = invoiceJournal.split("\r\n".toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray().size
            val pageSize = Rectangle(174f, (163 + rows * 7.7).toFloat())

            val document = Document(pageSize)
            document.setMargins(5f, 5f, 0f, 0f)

            PdfWriter.getInstance(document, FileOutputStream(file.absoluteFile))

            document.open()
            document.add(table)

            // close document
            document.close()

            true
        } catch (e: IOException) {
            e.printStackTrace()
            false
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }
}
*/
